### ISSUE TITLE
Same results for different queries using whereRaw with added bindings

### ISSUE THREAD

**Issue by emielmolenaar** [GeneaLabs/laravel-model-caching#111]
# Issue
I get the same results for different queries using whereRaw with added bindings when called on the same column. For example:

```
SomeModel::whereRaw('somecolumn = ?', ['somevalue'])->get();
SomeModel::whereRaw('somecolumn = ?', ['anothervalue'])->get();
SomeModel::whereRaw('somecolumn = ?', ['yetanothervalue'])->get();
```
The second and third query always returns the result of the first query.

I did dug around in the code. I noticed that `CacheKey` does not include the value of the binding given to the `whereRaw` method. This results in a string that is the same for these three queries, thus causing trouble.

I guess adding a `$this->getValuesClause($where)` to https://github.com/GeneaLabs/laravel-model-caching/blob/master/src/CacheKey.php#L172 should fix the problem? Haven't been able to spend much time on it so please correct me if I'm wrong or if this will break stuff. I am willing to create a PR to fix this!

## Environment
Laravel Version: 5.6.16
Laravel Model Caching Package Version: 0.2.52
PHP Version: 7.2.2
Homestead Version: n/a
Operating System & Version: Debian 9.3

**Comment by mikebronner** [comment]
Thanks for submitting this! :) I will take a look at this as soon as I can. :)

**Comment by mikebronner** [comment]
Please try release 0.2.54 or greater -- that should have fixed it. Please let me know. :) And thanks for being patient!

**Comment by emielmolenaar** [comment]
Thanks @mikebronner ! Handling raws is going better now, except when they are in a nested where :smiley: 

```
ErrorException: Undefined offset: 6 in vendor/genealabs/laravel-model-caching/src/CacheKey.php:192

#2 vendor/genealabs/laravel-model-caching/src/CacheKey.php(152): GeneaLabs\LaravelModelCaching\CacheKey->getRawClauses(Array)
#3 [internal function]: GeneaLabs\LaravelModelCaching\CacheKey->GeneaLabs\LaravelModelCaching\{closure}('-_or_LOWER(`ass...', Array)
#4 vendor/laravel/framework/src/Illuminate/Support/Collection.php(1313): array_reduce(Array, Object(Closure), NULL)
#5 vendor/genealabs/laravel-model-caching/src/CacheKey.php(156): Illuminate\Support\Collection->reduce(Object(Closure))
#6 vendor/genealabs/laravel-model-caching/src/CacheKey.php(167): GeneaLabs\LaravelModelCaching\CacheKey->getWhereClauses(Array)
#7 vendor/genealabs/laravel-model-caching/src/CacheKey.php(150): GeneaLabs\LaravelModelCaching\CacheKey->getNestedClauses(Array)
#8 [internal function]: GeneaLabs\LaravelModelCaching\CacheKey->GeneaLabs\LaravelModelCaching\{closure}(NULL, Array)
#9 vendor/laravel/framework/src/Illuminate/Support/Collection.php(1313): array_reduce(Array, Object(Closure), NULL)
#10 vendor/genealabs/laravel-model-caching/src/CacheKey.php(156): Illuminate\Support\Collection->reduce(Object(Closure))
#11 vendor/genealabs/laravel-model-caching/src/CacheKey.php(36): GeneaLabs\LaravelModelCaching\CacheKey->getWhereClauses()
#12 vendor/genealabs/laravel-model-caching/src/Traits/Caching.php(75): GeneaLabs\LaravelModelCaching\CacheKey->make(Array, NULL, '')
#13 vendor/genealabs/laravel-model-caching/src/CachedBuilder.php(77): GeneaLabs\LaravelModelCaching\CachedBuilder->makeCacheKey(Array)

```

I can reproduce this issue this way:

```
SomeModel::select(['column1', 'column2', 'column3', 'column4', 'column5', 'column6'])->where(function($where) { 
  $where->orWhereRaw('column1 LIKE ?', [$somevar]);
  $where->orWhereRaw('column2 LIKE ?', [$somevar]);
  $where->orWhereRaw('column3 LIKE ?', [$somevar]);
  $where->orWhereRaw('column4 LIKE ?', [$somevar]);
  $where->orWhereRaw('column5 LIKE ?', [$somevar]);
  $where->orWhereRaw('column6 LIKE ?', [$somevar]);
});
```
I don't think `$this->currentBinding++;` is required in `getNestedClauses()` ?

**Comment by mikebronner** [comment]
@Marcuzz can you check if this was fixed in 0.2.55 ... there were some issues I fixed from 0.2.54. Thanks!

**Comment by Marcuzz** [comment]
@mikebronner It is not, it's still an issue in 0.2.55 as that's the version I tested this on, thanks

**Comment by mikebronner** [comment]
Thanks, will take a look -- I know there were other issues with nested where clauses. Will try to fix both over the next few days.

**Comment by mikebronner** [comment]
@emielmolenaar I am unable to recreate the issue. Running `select()` on a model has no effect, it's really only useful on `querybuilders`, and not on `eloquentquerybuilders`.

The code you proveded above returns an instance of eloquentquerybuilder, and not a a result as such. Could you provide a working example that reproduces the issue? That would help tremendously! Thanks!

**Comment by mikebronner** [comment]
@emielmolenaar I'll close this for now. Please re-open if this problem persists.

**Comment by emielmolenaar** [comment]
Hi @mikebronner , thanks for your patience :smiley: A little background information: I ran into the issue when using this package in combination with https://github.com/yajra/laravel-datatables . When searching a table, the laravel-datatables package performs a whereRaw search under the hood. 

The reason why I'm using `select` on an EloquentQueryBuilder is to speed up the whole process of querying: we are using tables with a lot of columns which also produce ambigious column names in the case of joining tables together.

I have just updated all packages to their latest version on the project where I ran into the issue and I can not reproduce it anymore. I am not getting the ErrorException about an undefined offset. I guess the work you have done in the meantime involving src/CacheKey.php has fixed the notice.

TLDR; fixed! :tada:

**Comment by mikebronner** [comment]
Good news! :) Thanks for following up.

**Comment by rappasoft** [comment]
I too still have that problem using laravel-datatables while caching is on. I'm using 0.2.62 currently, and it only happens for me when using the datatables search.

The error is as follows: 

`ErrorException: Undefined offset: 3 in PROJECT_PATH/vendor/genealabs/laravel-model-caching/src/CacheKey.php:235`

If I disable model caching the search works as intended.

Gonna dig around but i'm not sure where to really begin.

**Comment by mikebronner** [comment]
@rappasoft Could you please open a new issue and post the complete stack trace, the eloquent query which causes the error, and anything else that you think might be pertinent to the problem. Thanks!

**Comment by rappasoft** [comment]
@mikebronner Yes of course trying to gather as much insight as I can.

**Comment by jf-m** [comment]
Hi @mikebronner ,

I'm currently having the same bug as the OP here with the latest version.
```
SomeModel::whereRaw('somecolumn = ?', ['somevalue'])->get();
SomeModel::whereRaw('somecolumn = ?', ['anothervalue'])->get();
SomeModel::whereRaw('somecolumn = ?', ['yetanothervalue'])->get();
```
The second and third query always returns the result of the first query.


I guess that regarding your comment https://github.com/GeneaLabs/laravel-model-caching/issues/111#issuecomment-386921683 it was supposed to be fixed, hope I'm right.

Thank you and cheers !

**Comment by mikebronner** [comment]
@jf-m Could you test with version 0.4.16 and open a new issue if the problem persists? Thanks!
